categories-and-dynamic-typing/dyn-cats.tex

```diff
@@ -0,0 +1,696 @@
+  \end{theorem}
+  That is, from a model of untyped lambda calculus, we can
+  \emph{construct} a cartesian closed category with a reflexive object
+  such that the functions in the untyped model are exactly the
+  endo-functions of the reflexive object.
+
+  Remembering the correspondence between typed lambda calculus and
+  cartesian closed categories, \emph{constructing} a CCC is the same
+  as defining a \emph{translation} of the typed lambda calculus, so
+  the proof of this theorem is actually a translation from typed to
+  untyped lambda calculus.
+
+  The intuition for the translation comes from the previous section:
+  we want the types to be retracts of the dynamic type, so we will
+  interpret the types as exactly the idempotent they represent on the
+  dynamic type, which is then just an untyped lambda term!
+
+  So first, we interpret the types of the lambda calculus as untyped
+  terms $c$ such that $c(c x) \equiv c x$.
+
+  \begin{mathpar}
+    \begin{array}{rcl}
+      \compile{1} & = & \lambda x.()\\
+      \compile{A\times B} & = & \lambda p. (\compile{A}(\pi_1 p), \compile{B}(\pi_2 p))\\
+      \compile{A\to B} & = & \lambda f. \lambda x. \compile{B}(f(\compile{A} x))
+    \end{array}
+  \end{mathpar}
+
+  Which you may recognize as (except the $1$) case, basically the
+  corresponding contracts from \cite{findler2002contracts}. I'll leave
+  it as an exercise that these are actually idempotent.
+
+  Next, we want to translate typed terms to untyped terms, but in such
+  a way that we maintain the validity of $\beta,\eta$ equivalence. How
+  can we achieve this? When you expand the domain of a function, you
+  can call it on new inputs, so we have to make sure that these new
+  inputs don't reveal any new information about our terms. That is,
+  the equality for the untyped translation of a typed term should be
+  completely determined by its behavior linking with other typed
+  terms.
+
+  Fortunately, the interpretation of types above coerces \emph{any}
+  untyped term to behave like a term of a given type. So we just need
+  to make sure that any interaction with a dynamically typed term is
+  mediated by one of these idempotents.
+
+  It turns out that it's very simple to
+  derive a translation that validates this equation, we just need to
+  put idempotents on the variables!
+
+  \begin{mathpar}
+    \begin{array}{rcl}
+      \compile{x : A} & = & \compile{A} x\\
+      \compile{\lambda (x:A). t}  & = & \lambda x. \compile{t}\\
+      \compile{t u}  & = & \compile{t}\compile{u}\\
+      \compile{(t,u)}  & = & (\compile{t},\compile{u})\\
+      \compile{\pi_i t}  & = & \pi_i \compile{t}\\
+      \compile{()} & = & ()
+    \end{array}
+  \end{mathpar}
+
+  In order to prove preservation of $\eta$, we prove the following
+  soundness theorem.
+
+  \begin{theorem}[Soundness]
+    \label{theorem:soundness}
+    {~}\\
+    If $x_1:A_1,\ldots,x_n:A_n\vdash{} t : B$, then
+    $x_1,\ldots,x_n\vdash{} \compile{t}$ and
```

Comment:
not sure what the second judgment means (`\Gamma \vdash \compile{t}`). Is this saying `\compile{t}` is well-formed?